tests/kernel/mem_protect/obj_validation/src/main.c

```diff
@@ -112,10 +117,101 @@ void test_generic_object(void)
 	}
 }
 
+/**
+ * @brief Test requestor thread will implicitly be assigned permission on the
+ * dynamically allocated object
+ *
+ * @details
+ * - Create kernel object semaphore, dynamically allocate it from the calling
+ *   thread's resource pool.
+ * - Check that object's address is in bounds of that memory pool.
+ * - Then check the requestor thread will implicitly be assigned permission on
+ *   the allocated object by using semaphore API k_sem_init()
+ *
+ * @ingroup kernel_memprotect_tests
+ *
+ * @see k_object_alloc()
+ */
+void test_kobj_assign_perms_on_alloc_obj(void)
+{
+	test_dyn_sem0 = k_object_alloc(K_OBJ_SEM);
+	zassert_not_null(test_dyn_sem0, "Cannot allocate sem k_object");
+
+	struct k_thread *thread = _current;
+
+	intptr_t sem_addr = (intptr_t)&test_dyn_sem0;
+	intptr_t start_addr = (intptr_t)&thread->resource_pool;
+	size_t size_heap = CONFIG_HEAP_MEM_POOL_SIZE;
+
+	/* check that kernel object may be dynamically allocated,
+	 * drawing memory from the calling thread's memory pool
+	 */
+#if defined(CONFIG_X86) || (CONFIG_ARM)
+	zassert_true(
+		(sem_addr > start_addr) && (sem_addr <= (start_addr + size_heap)),
+		"semaphore not in bound of thread's memory pool");
+#elif defined(CONFIG_ARC)
+	zassert_true(
+		(sem_addr < start_addr) && (sem_addr > (start_addr - size_heap)),
```

Comment:
Sorry it's not immediately clear to me why for ARC there's a special treatment (I'm not saying it's wrong but so far I don't understand how ARC is different from ARM & x86).

@ruuddw any guesses?

Reply:
this is also unclear to me.
Why is arch-specific logic needed here?

Reply:
No idea, other then masking some problem with the code. The check does not seem to check what is claim to check: &test_dyn_sem0 gives an address on the stack, not an address in the calling thread's memory pool.

Reply:
@ruuddw I thought that stack in ARC growing in other direction, so in fact it was just make code works... I will rewrite that test.

Reply:
@maksimmasalski on ARC stack grows from higher addresses to lower addresses, see what Programmer's Reference Manual (you may get your copy for free here after a click-through registration: https://www.synopsys.com/dw/doc.php/ds/cc/programmers-reference-manual-ARC-EM.pdf) says about `push_p` instruction for example:
> PUSH_S
>
> Function: Push onto Stack
> 
> Operation
> SP=SP-4; *SP = b;

Reply:
> @maksimmasalski on ARC stack grows from higher addresses to lower addresses, see what Programmer's Reference Manual (you may get your copy for free here after a click-through registration: [synopsys.com/dw/doc.php/ds/cc/programmers-reference-manual-ARC-EM.pdf](https://www.synopsys.com/dw/doc.php/ds/cc/programmers-reference-manual-ARC-EM.pdf)) says about `push_p` instruction for example:
> 
> > PUSH_S
> > Function: Push onto Stack
> > Operation
> > SP=SP-4; *SP = b;

Thank you, Alexey. Downloaded it)

Reply:
@abrodkin @ruuddw @andrewboie I removed all arch specific code, because it was unnecessary.

Reply:
> > @maksimmasalski on ARC stack grows from higher addresses to lower addresses, see what Programmer's Reference Manual (you may get your copy for free here after a click-through registration: [synopsys.com/dw/doc.php/ds/cc/programmers-reference-manual-ARC-EM.pdf](https://www.synopsys.com/dw/doc.php/ds/cc/programmers-reference-manual-ARC-EM.pdf)) says about `push_p` instruction for example:
> > > PUSH_S
> > > Function: Push onto Stack
> > > Operation
> > > SP=SP-4; *SP = b;
> 
> Thank you, Alexey. Downloaded it)

Note even though both ARC EM & ARC HS implement the same generation of ARC ISA (we call it ARCv2 ISA) there're still some differences in supported instructions and sometimes even in their semantics, that said go get ARC HS PRM here for completeness: https://www.synopsys.com/dw/doc.php/ds/cc/programmers-reference-manual-ARC-HS.pdf

Reply:
> I removed all arch specific code, because it was unnecessary.

Thanks, this looks right now: checking the k_object pointer instead of a stack pointer.

Reply:
> > I removed all arch specific code, because it was unnecessary.
> 
> Thanks, this looks right now: checking the k_object pointer instead of a stack pointer.

Thank you. Do you think other two tests `test_krnl_obj_static_alloc_build_time()` and `test_no_ref_dyn_kobj_release_mem()` are good enough?